storage-proofs-update/src/circuit.rs

```diff
@@ -0,0 +1,587 @@
+}
+
+impl PublicInputs {
+    // Returns public-inputs in the order expected by `ConstraintSystem::verify()`.
+    pub fn to_vec(&self) -> Vec<Fr> {
+        vec![
+            Fr::from(self.k as u64),
+            self.comm_r_old.into(),
+            self.comm_d_new.into(),
+            self.comm_r_new.into(),
+            Fr::from(self.h_select),
+        ]
+    }
+}
+
+pub struct ChallengeProof<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    pub leaf_r_old: Option<Fr>,
+    pub path_r_old: Vec<Vec<Option<Fr>>>,
+    pub leaf_d_new: Option<Fr>,
+    pub path_d_new: Vec<Vec<Option<Fr>>>,
+    pub leaf_r_new: Option<Fr>,
+    pub path_r_new: Vec<Vec<Option<Fr>>>,
+    pub _tree_r: PhantomData<TreeR>,
+}
+
+// Implement `Clone` by hand because `MerkleTreeTrait` does not implement `Clone`.
+impl<TreeR> Clone for ChallengeProof<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    fn clone(&self) -> Self {
+        ChallengeProof {
+            leaf_r_old: self.leaf_r_old,
+            path_r_old: self.path_r_old.clone(),
+            leaf_d_new: self.leaf_d_new,
+            path_d_new: self.path_d_new.clone(),
+            leaf_r_new: self.leaf_r_new,
+            path_r_new: self.path_r_new.clone(),
+            _tree_r: PhantomData,
+        }
+    }
+}
+
+impl<TreeR> ChallengeProof<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    pub fn from_merkle_proofs(
+        proof_r_old: MerkleProof<
+            TreeRHasher,
+            TreeR::Arity,
+            TreeR::SubTreeArity,
+            TreeR::TopTreeArity,
+        >,
+        proof_d_new: MerkleProof<TreeDHasher, TreeDArity>,
+        proof_r_new: MerkleProof<
+            TreeRHasher,
+            TreeR::Arity,
+            TreeR::SubTreeArity,
+            TreeR::TopTreeArity,
+        >,
+    ) -> Self {
+        let leaf_r_old = Some(proof_r_old.leaf().into());
+        let path_r_old: Vec<Vec<Option<Fr>>> = proof_r_old
+            .path()
+            .iter()
+            .map(|(siblings, _insert)| siblings.iter().map(|&s| Some(s.into())).collect())
+            .collect();
+
+        let leaf_d_new = Some(proof_d_new.leaf().into());
+        let path_d_new: Vec<Vec<Option<Fr>>> = proof_d_new
+            .path()
+            .iter()
+            .map(|(siblings, _insert)| siblings.iter().map(|&s| Some(s.into())).collect())
+            .collect();
+
+        let leaf_r_new = Some(proof_r_new.leaf().into());
+        let path_r_new: Vec<Vec<Option<Fr>>> = proof_r_new
+            .path()
+            .iter()
+            .map(|(siblings, _insert)| siblings.iter().map(|&s| Some(s.into())).collect())
+            .collect();
+
+        ChallengeProof {
+            leaf_r_old,
+            path_r_old,
+            leaf_d_new,
+            path_d_new,
+            leaf_r_new,
+            path_r_new,
+            _tree_r: PhantomData,
+        }
+    }
+
+    pub fn blank(sector_nodes: usize) -> Self {
```

Comment:
I think we use `empty`  in the other circuits, can we keep consistent with that?

Reply:
Yup, I can do that. I chose `blank` because its called by `EmptySectorUpdateCompound::blank_circuit`.